app/services/CartServiceImpl.java

```diff
@@ -0,0 +1,156 @@
+        final AddLineItem action = AddLineItem.of(product.getId(), variant.getId(), frequency);
+        final Cart updatedCart = sphereClient().execute(CartUpdateCommand.of(clearedCart, action)).toCompletableFuture().join();
+
+        final CustomObjectDraft<Integer> draft = CustomObjectDraft.ofUnversionedUpsert(FREQUENCY, updatedCart.getId(), frequency,
+                new TypeReference<CustomObject<Integer>>() {
+                });
+
+        final CustomObject<Integer> customObject = sphereClient().execute(CustomObjectUpsertCommand.of(draft)).toCompletableFuture().join();
+        Logger.debug("Setting new or update CustomObject: {}", customObject);
+    }
+
+    @Override
+    public int getFrequency(final String cartId) {
+        final Optional<CustomObject<JsonNode>> result = Optional.ofNullable(
+                sphereClient().execute(CustomObjectByKeyGet.of(FREQUENCY, cartId)).toCompletableFuture().join());
+        if (result.isPresent()) {
+            return result.get().getValue().asInt();
+        }
+        return 0;
+    }
+
+    @Override
+    public Optional<ProductVariant> getSelectedVariant(final Cart cart) {
+        final Optional<ProductVariant> selectedVariant =
+                (cart.getLineItems().size() > 0)
+                        ? Optional.ofNullable(cart.getLineItems().get(0).getVariant())
+                        : Optional.empty();
+        return selectedVariant;
+    }
+
+    @Override
+    public Cart createCartWithPactasInfo(final ProductProjection product, final PactasContract contract, final PactasCustomer customer) {
+        final ProductVariant variant = getVariantInContract(product, contract);
+        final Cart cart = sphereClient().execute(CartCreateCommand.of(CartDraft.of(DefaultCurrencyUnits.EUR))).toCompletableFuture().join();
+        Logger.debug("Created new Cart[cartId={}] with Pactas info", cart.getId());
+
+        final AddLineItem action = AddLineItem.of(product.getId(), variant.getId(), 1);
+        final Cart updatedCart = sphereClient().execute(CartUpdateCommand.of(cart, action)).toCompletableFuture().join();
+
+        final Address address = AddressBuilder.of(customer.getCompleteAddress()).build();
+        final Cart cartWithAddress = sphereClient().execute(CartUpdateCommand.of(updatedCart, SetShippingAddress.of(address))).toCompletableFuture().join();
+        return cartWithAddress;
+    }
+
+    private ProductVariant getVariantInContract(final ProductProjection product, final PactasContract contract) {
+        final String planVariantId = contract.getPlanVariantId();
+        final Optional<ProductVariant> variant = variant(product, planVariantId);
+        if (variant.isPresent()) {
+            return variant.get();
+        }
+        throw new PlanVariantNotFound(planVariantId);
+    }
+
+    private Optional<ProductVariant> variant(final ProductProjection product, final String pactasId) {
+        for (final ProductVariant variant : product.getAllVariants()) {
```

Comment:
This method has exit points in a loop, can you rewrite it using `product.getAllVariants().stream()`?